### PR TITLE
Fix rustdoc ICE on macros defined within functions

### LINKED ISSUE #47639: ICE building docs of some crates in beta
When building the documentation (or running the doctests) of a few crates, such as libsecp256k1, in beta or nightly, rustdoc crashes with this message (while it works fine on stable, even if with some pulldown warnings):

```
thread 'rustc' panicked at '/checkout/src/librustc/lint/context.rs:1049: failed to process bufferd lint here', /checkout/src/librustc/session/mod.rs:1130:26
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at /checkout/src/libstd/sys_common/backtrace.rs:68
             at /checkout/src/libstd/sys_common/backtrace.rs:57
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:391
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
   6: rustc::ty::context::tls::with_opt
   7: rustc::session::span_bug_fmt
   8: rustc::lint::context::check_ast_crate
   9: rustc_driver::driver::phase_2_configure_and_expand
             at /checkout/src/librustc_driver/driver.rs:909
  10: rustdoc::core::run_core
             at /checkout/src/librustdoc/core.rs:161

error: Could not document `libsecp256k1`.
```

* Crate [libsecp256k1](https://crates.io/crates/libsecp256k1) regressed from stable to beta ([crater log caused by doctests](http://cargobomb-reports.s3.amazonaws.com/1.23.0-1.24.0-beta.3/beta-2018-01-13/reg/libsecp256k1-0.1.12/log.txt)).
* Crate [oftlisp-anf](https://crates.io/crates/oftlisp-anf) regressed from stable to beta ([crater log caused by doctests](http://cargobomb-reports.s3.amazonaws.com/1.23.0-1.24.0-beta.3/beta-2018-01-13/reg/oftlisp-anf-0.1.3/log.txt)).

cc @sorpaas

### PR DESCRIPTION
**Pull request by Manishearth**
fixes #47639

### CODE REVIEW

File: src/test/rustdoc/issue-47639.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This should not ICE
```
Comment (kennytm) [inline src/test/rustdoc/issue-47639.rs]:
```
[00:05:40] tidy error: /checkout/src/test/rustdoc/issue-47639.rs:11: trailing whitespace
```

**Comment by Manishearth** [comment]
fixed

-Manish Goregaokar

**Comment by QuietMisdreavus** [comment]
Seeking a second opinion. Does someone on @rust-lang/compiler have a better grasp of the ramifications of turning off the "buffered-but-not-fired lints are bugs" functionality when running as rustdoc? @jseyfried? Something about "just patching out the line that panics" seems off, but then again the "everybody loops" pass is also a bit strange, so maybe it balances out.

**Comment by Manishearth** [comment]
The assumption made in that `bug!` is that all linted nodes must be visited. This is an invariant that holds true in the absence of the everybody-loops pass (and if it were to break it would be problematic), however everybody-loops throws things for a toss by removing nodes wholesale. Proc macros can also remove nodes, but not _after_ they get linted (admittedly a shaky way of holding up the invariant, but hey, that's why we have the `bug!`)

A more robust solution would be to note down the macros removed by everybody-loops but imo that's a bit like overkill.

**Comment by Mark-Simulacrum** [comment]
Could we instead gate this check on whether everybody loops pass was run? Not sure whether that's *better* necessarily, but it seems at least like a minor improvement.

**Comment by Manishearth** [comment]
Where should I stash that fact? On the session? On the context?

One trick that could be done is to empty that buffer immediately after
running the pass, or turn off the unused macro built-in for rustdoc.

**Comment by Mark-Simulacrum** [comment]
Hm. Yeah, let's not change anything then unless the compiler team decides on something. Is there a chance you could prepare a PR to backport this (minimal) fix to beta? It seems like the smallest change we can make that should fix this regression and I'd like to get it on beta ASAP.

**Comment by Manishearth** [comment]
done: https://github.com/rust-lang/rust/pull/47969

won't be around for a while, feel free to force push to that branch or whatever to get it landed

File: src/librustc/lint/context.rs [hunk 2]
```diff
@@ -1042,11 +1042,20 @@ pub fn check_ast_crate(sess: &Session, krate: &ast::Crate) {
...
+    // All of the buffered lints should have been emitted at this point.
+    // If not, that means that we somehow buffered a lint for a node id
+    // that was not lint-checked (perhaps it doesn't exist?). This is a bug.
+    //
+    // Rustdoc runs everybody-loops before the early lints and removes
+    // function bodies, so it's totally possible for linted
+    // node ids to not exist (e.g. macros defined within functions for the
+    // unused_macro lint) anymore. So we only run this check
+    // when we're not in rustdoc mode. (see issue #47639)
+    if !sess.opts.actually_rustdoc {
```
Comment (eddyb) [inline src/librustc/lint/context.rs]: `actually_rustdoc` is pre-existing?!

Comment (eddyb) [reply src/librustc/lint/context.rs]: Oh this is not CLI-facing, just an internal flag used for literally causing the problem here.

**Comment by Manishearth** [comment]
Anyway, r?

We've landed the beta one so this isn't urgent anymore but nightly is still broken.

**Comment by Mark-Simulacrum** [comment]
I think we should land this and perhaps iterate later if necessary. If anyone objects, feel free to r-

@bors r+

**Pull request merged**